Rework

AmjadHD opened the pull request:
Fix #155
Fix #152
Fix #151
Fix #150
Fix #149
Fix #147
Fix #145
Fix #144
Fix #143
Fix #139
Fix #90
Fix #56
Fix #43

Araq commented:
@Varriount Ping.

Varriount commented on Syntaxes/Nim.YAML-tmLanguage:
```diff
@@ -2,572 +2,985 @@
+
+  invalid-names:
+    patterns:
+      - match: '(?x:addr|asm|bind|block|break|case|cast|concept|const|continue|converter
+        |defer|discard|distinct|do|elif|else|end|enum|except|export|finally|for
+        |from|func|if|import|include|interface|iterator|let|macro|method|mixin
+        |object|of|out|proc|ptr|raise|ref|return|static|template|try|tuple|type
+        |using|var|when|while|yield)\b'
```
Because this group is marked by the `x` flag, I believe it can be reformatted:
```suggestion
      - match: |
          (?x:
            addr      |
            asm       |
            bind      |
            block     |
            break     |
            case      |
            cast      |
            concept   |
            const     |
            continue  |
            converter |
            defer     |
            discard   |
            distinct  |
            do        |
            elif      |
            else      |
            end       |
            enum      |
            except    |
            export    |
            finally   |
            for       |
            from      |
            func      |
            if        |
            import    |
            include   |
            interface |
            iterator  |
            let       |
            macro     |
            method    |
            mixin     |
            object    |
            of        |
            out       |
            proc      |
            ptr       |
            raise     |
            ref       |
            return    |
            static    |
            template  |
            try       |
            tuple     |
            type      |
            using     |
            var       |
            when      |
            while     |
            yield
          )\b
```

AmjadHD replied:
But why ?

AmjadHD commented:
https://github.com/nim-lang/Nim/blob/devel/doc/mytest.cfg#L4
Is this valid syntax in nim cfg files ?

Varriount commented on Syntaxes/Nim CFG.sbnf:
```diff
@@ -0,0 +1,218 @@
+EXTENSIONS = 'nim.cfg'
```
Is this file meant to be used with any specific tool, or is it just for reference?

AmjadHD replied:
https://github.com/BenjaminSchaaf/sbnf

Varriount replied:
Hm, I'm going to have to think on this. I'm not sure about adding another dependency to this package's development.

Varriount replied:
Hm, I'm going to have to think on this. I'm not sure about adding another dependency to this package's development.

AmjadHD commented on Syntaxes/Nim.YAML-tmLanguage:
```diff
@@ -2,572 +2,985 @@
+        name: meta.function.nim
+        patterns:
+          - include: '#comments'
+          - include: '#name'
```
@Varriount I'm not sure how to make this match at most one symbol.

Varriount commented:
One thing that needs to be kept in mind here is that [Linguist](https://github.com/github/linguist) (GitHub's code highlighter) actually uses [NimLime's syntax definitions](https://github.com/github/linguist/blob/987bb6708f9ffc5658a1605ccdad6b4978915e4e/grammars.yml#L64) for code highlighting. Last I checked, Linguist only works with TextMate-compatible grammers, so it doesn't support Sublime Text's additional features.

AmjadHD commented:
@Varriount, review pls.

AmjadHD commented:
Is this going to be reviewed any time soon ?

Araq commented:
Ping @Varriount

AmjadHD commented:
@Varriount, There is a new ST plugin for Nim [NimPlus](https://github.com/vanyle/NimPlus), should I just add the new `.sublime-syntax` files there ?

AmjadHD commented:
@Varriount, The way forward as discussed is to split this repo to two branches one that contains old `.tmLanguage` files for github and another that contains `.sublime-syntax` files for Sublime Text (and Sublime Merge), so that the testing of each format is seamless.
> Can you give me an overview of what has changed?

It's a rework so many things have changed, you can try it and observe the differences for yourself.
Otherwise, off the top of my head:
* Update to recent Nim, strings, numbers, chars fixed and tested, `fmt` and `&` strings (only for `fmt"` not for `fmt "` or any other form), better call syntax, highlighting of parameters (proc calls and definitions, tuples and generics), types, const, var, ... sections highlighted (for top level and 2-space indentations only), do syntax, the syntax is faster etc.
* reworked Nim config file using sbnf.
* reworked Nim filter syntax 
* simplified indentation rules.
* removed some unused files.

lildude commented:
👋 from [github/linguist](https://github.com/github/linguist). This PR has introduced a bug our compiler has just detected whilst I was preparing for the next release:

```
- [ ] repository `vendor/grammars/NimLime` (from https://github.com/nim-lang/NimLime) (1 errors)
  - Grammar conversion failed. File `Syntaxes/Nim CFG.tmLanguage` failed to parse: 1 error(s) decoding:

* 'Repository[block-comments]' expected a map, got 'slice'
```

Varriount commented:
@lildude It should be fixed now.

lildude commented:
Confirmed. Thanks.
